File: clients/python_pypy/debugging/debug_state.py
from debugging.camera import Camera
from model.vec2 import Vec2
from stream_wrapper import StreamWrapper
from typing import List
from typing import Optional

class DebugState:
    """Renderer's state"""

    __slots__ = ("pressed_keys","cursor_world_position","locked_unit","camera",)

    pressed_keys: List[str]
    cursor_world_position: Vec2
    locked_unit: Optional[int]
    camera: Camera

    def __init__(self, pressed_keys: List[str], cursor_world_position: Vec2, locked_unit: Optional[int], camera: Camera):
        self.pressed_keys = pressed_keys
        """Pressed keys"""
        self.cursor_world_position = cursor_world_position
        """Cursor's position in game coordinates"""
        self.locked_unit = locked_unit
        """Id of unit which is followed by the camera, or None"""
        self.camera = camera
        """Current camera state"""

    @staticmethod
    def read_from(stream: StreamWrapper) -> "DebugState":
        """Read DebugState from input stream
        """
        pressed_keys = []
        for _ in range(stream.read_int()):
            pressed_keys_element = stream.read_string()
            pressed_keys.append(pressed_keys_element)
        cursor_world_position = Vec2.read_from(stream)
        if stream.read_bool():
            locked_unit = stream.read_int()
        else:
            locked_unit = None
        camera = Camera.read_from(stream)
        return DebugState(pressed_keys, cursor_world_position, locked_unit, camera)
    
    def write_to(self, stream: StreamWrapper):
        """Write DebugState to output stream
        """
        stream.write_int(len(self.pressed_keys))
        for element in self.pressed_keys:
            stream.write_string(element)
        self.cursor_world_position.write_to(stream)
        if self.locked_unit is None:
            stream.write_bool(False)
        else:
            stream.write_bool(True)
            stream.write_int(self.locked_unit)
        self.camera.write_to(stream)
    
    def __repr__(self):
        return "DebugState(" + \
            repr(self.pressed_keys) + \
            ", " + \
            repr(self.cursor_world_position) + \
            ", " + \
            repr(self.locked_unit) + \
            ", " + \
            repr(self.camera) + \
            ")"

File: clients/swift/Sources/AiCup22/Debugging/DebugState.swift
/// Renderer's state
public struct DebugState {
    /// Pressed keys
    let pressedKeys: [String]

    /// Cursor's position in game coordinates
    let cursorWorldPosition: Vec2

    /// Id of unit which is followed by the camera, or None
    let lockedUnit: Int32?

    /// Current camera state
    let camera: Camera

    /// Read DebugState from input stream
    static func readFrom<S: InputStream>(_ stream: S) -> DebugState {
        var pressedKeys: [String]
        let pressedKeysSize = stream.readInt32()
        pressedKeys = (0..<pressedKeysSize).map{ _ in
            var pressedKeysSize: String
            pressedKeysSize = stream.readString()
            return pressedKeysSize
        }
        var cursorWorldPosition: Vec2
        cursorWorldPosition = Vec2.readFrom(stream)
        var lockedUnit: Int32?
        if stream.readBool() {
            lockedUnit = stream.readInt32()
        } else {
            lockedUnit = nil
        }
        var camera: Camera
        camera = Camera.readFrom(stream)
        return DebugState(pressedKeys: pressedKeys, cursorWorldPosition: cursorWorldPosition, lockedUnit: lockedUnit, camera: camera)
    }

    /// Write DebugState to output stream
    func writeTo<S: OutputStream>(_ stream: S) {
        stream.writeInt32(Int32(pressedKeys.count))
        for pressedKeysElement in pressedKeys {
            stream.writeString(pressedKeysElement)
        }
        cursorWorldPosition.writeTo(stream)
        if lockedUnit == nil {
            stream.writeBool(false)
        } else {
            stream.writeBool(true)
            let lockedUnitValue = lockedUnit!
            stream.writeInt32(lockedUnitValue)
        }
        camera.writeTo(stream)
    }
}
